File: src/test/java/funding/startreum/domain/virtualaccount/service/AccountChargeServiceTest.java
package funding.startreum.domain.virtualaccount.service;

import funding.startreum.domain.transaction.entity.Transaction;
import funding.startreum.domain.transaction.service.TransactionService;
import funding.startreum.domain.virtualaccount.dto.request.AccountRequest;
import funding.startreum.domain.virtualaccount.dto.response.AccountPaymentResponse;
import funding.startreum.domain.virtualaccount.entity.VirtualAccount;
import funding.startreum.domain.virtualaccount.repository.VirtualAccountRepository;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.bean.override.mockito.MockitoBean;

import java.math.BigDecimal;
import java.time.LocalDateTime;

import static funding.startreum.domain.transaction.entity.Transaction.TransactionType.REMITTANCE;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.ArgumentMatchers.isNull;
import static org.mockito.Mockito.when;

@SpringBootTest
class AccountChargeServiceTest {

    @MockitoBean
    private VirtualAccountRepository virtualAccountRepository;

    @MockitoBean
    private TransactionService transactionService;

    @MockitoBean
    private AccountQueryService accountQueryService;

    @Autowired
    private AccountChargeService accountChargeService;

    @Nested
    @DisplayName("chargeByAccountId() 테스트")
    class ChargeByAccountIdTests {
        @Test
        void testChargeByAccountId() {
            int accountId = 1;
            BigDecimal initialBalance = BigDecimal.valueOf(100);
            BigDecimal chargeAmount = BigDecimal.valueOf(50);
            AccountRequest request = new AccountRequest(chargeAmount);

            // 테스트용 VirtualAccount 생성
            VirtualAccount account = new VirtualAccount();
            account.setAccountId(accountId);
            account.setBalance(initialBalance);

            when(accountQueryService.getAccount(accountId)).thenReturn(account);

            // 거래 생성 모의 (첫번째 파라미터 null)
            Transaction transaction = new Transaction();
            transaction.setTransactionId(1);
            LocalDateTime now = LocalDateTime.now();
            transaction.setTransactionDate(now);
            when(transactionService.createTransaction(isNull(), eq(account), eq(account), eq(chargeAmount), eq(REMITTANCE)))
                    .thenReturn(transaction);

            AccountPaymentResponse response = accountChargeService.chargeByAccountId(accountId, request);

            // 충전 후 잔액은 초기 잔액 + 충전 금액이어야 함
            assertEquals(initialBalance.add(chargeAmount), account.getBalance());
            assertEquals(transaction.getTransactionId(), response.transactionId());
            assertEquals(accountId, response.accountId());
            assertEquals(initialBalance, response.beforeMoney());
            assertEquals(chargeAmount, response.chargeAmount());
            assertEquals(account.getBalance(), response.afterMoney());
            assertEquals(now, response.transactionDate());
        }
    }

    @Nested
    @DisplayName("chargeByUsername() 테스트")
    class ChargeByUsernameTests {
        @Test
        void testChargeByUsername() {
            String username = "testUser";
            BigDecimal initialBalance = BigDecimal.valueOf(200);
            BigDecimal chargeAmount = BigDecimal.valueOf(30);
            AccountRequest request = new AccountRequest(chargeAmount);

            VirtualAccount account = new VirtualAccount();
            account.setAccountId(2);
            account.setBalance(initialBalance);

            when(accountQueryService.getAccount(username)).thenReturn(account);

            Transaction transaction = new Transaction();
            transaction.setTransactionId(2);
            LocalDateTime now = LocalDateTime.now();
            transaction.setTransactionDate(now);
            when(transactionService.createTransaction(isNull(), eq(account), eq(account), eq(chargeAmount), eq(REMITTANCE)))
                    .thenReturn(transaction);

            AccountPaymentResponse response = accountChargeService.chargeByUsername(username, request);

            assertEquals(initialBalance.add(chargeAmount), account.getBalance());
            assertEquals(transaction.getTransactionId(), response.transactionId());
            assertEquals(account.getAccountId(), response.accountId());
            assertEquals(initialBalance, response.beforeMoney());
            assertEquals(chargeAmount, response.chargeAmount());
            assertEquals(account.getBalance(), response.afterMoney());
            assertEquals(now, response.transactionDate());
        }
    }
}

File: src/test/java/funding/startreum/domain/virtualaccount/service/AccountChargeServiceTest.kt
package funding.startreum.domain.virtualaccount.service

import funding.startreum.domain.transaction.entity.Transaction
import funding.startreum.domain.transaction.entity.Transaction.TransactionType
import funding.startreum.domain.transaction.service.TransactionService
import funding.startreum.domain.virtualaccount.dto.request.AccountRequest
import funding.startreum.domain.virtualaccount.entity.VirtualAccount
import funding.startreum.domain.virtualaccount.service.AccountChargeService
import funding.startreum.domain.virtualaccount.service.AccountQueryService
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.BDDMockito.given
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import java.math.BigDecimal
import java.time.LocalDateTime

@ExtendWith(MockitoExtension::class)
internal class AccountChargeServiceTest {

    @Mock
    lateinit var transactionService: TransactionService

    @Mock
    lateinit var accountQueryService: AccountQueryService

    @InjectMocks
    lateinit var accountChargeService: AccountChargeService

    lateinit var testAccount: VirtualAccount
    lateinit var testTransaction: Transaction
    lateinit var now: LocalDateTime

    @BeforeEach
    fun setUp() {
        testAccount = VirtualAccount().apply {
            accountId = 999
            balance = BigDecimal.valueOf(100)
        }
        now = LocalDateTime.now()
        testTransaction = Transaction().apply {
            transactionId = 9999
            transactionDate = now
        }
    }

    @Nested
    @DisplayName("chargeByAccountId() 테스트")
    inner class ChargeByAccountIdTests {
        @Test
        @DisplayName("정상 충전 시, 잔액 증가 및 Transaction 정보 확인")
        fun testChargeByAccountId() {
            // Given
            val accountId = 1
            val chargeAmount = BigDecimal.valueOf(50)
            val initialBalance = testAccount.balance
            val request = AccountRequest(chargeAmount)

            // 계좌 조회 시 가짜 계좌 반환
            given(accountQueryService.getAccountByAccountId(accountId)).willReturn(testAccount)

            given(
                transactionService.createTransaction(
                    null,
                    testAccount,
                    testAccount,
                    chargeAmount,
                    TransactionType.REMITTANCE
                )
            ).willReturn(testTransaction)

            // When
            val response = accountChargeService.chargeByAccountId(accountId, request)

            // Then
            // 잔액 증가 확인
            assertThat(testAccount.balance).isEqualTo(initialBalance.add(chargeAmount))
            // 응답 객체 검증
            assertThat(response.transactionId).isEqualTo(testTransaction.transactionId)
            assertThat(response.accountId).isEqualTo(testAccount.accountId)
            assertThat(response.beforeMoney).isEqualTo(initialBalance)
            assertThat(response.chargeAmount).isEqualTo(chargeAmount)
            assertThat(response.afterMoney).isEqualTo(testAccount.balance)
            assertThat(response.transactionDate).isEqualTo(now)
        }
    }

    @Nested
    @DisplayName("chargeByUsername() 테스트")
    inner class ChargeByUsernameTests {
        @Test
        @DisplayName("유저네임으로 충전 시, 잔액 증가 및 Transaction 정보 확인")
        fun testChargeByUsername() {
            // Given
            val username = "testUser"
            val chargeAmount = BigDecimal.valueOf(30)
            val initialBalance = testAccount.balance
            val request = AccountRequest(chargeAmount)

            // username 기반 가짜 계좌 반환
            given(accountQueryService.getAccountByUsername(username)).willReturn(testAccount)

            // transactionService.createTransaction(...) 호출 시 testTransaction 반환
            given(
                transactionService.createTransaction(
                    null,
                    testAccount,
                    testAccount,
                    chargeAmount,
                    TransactionType.REMITTANCE
                )
            ).willReturn(testTransaction)

            // When
            val response = accountChargeService.chargeByUsername(username, request)

            // Then
            // 잔액 증가 확인
            assertThat(testAccount.balance).isEqualTo(initialBalance.add(chargeAmount))
            // 응답 객체 검증
            assertThat(response.transactionId).isEqualTo(testTransaction.transactionId)
            assertThat(response.accountId).isEqualTo(testAccount.accountId)
            assertThat(response.beforeMoney).isEqualTo(initialBalance)
            assertThat(response.chargeAmount).isEqualTo(chargeAmount)
            assertThat(response.afterMoney).isEqualTo(testAccount.balance)
            assertThat(response.transactionDate).isEqualTo(now)
        }
    }
}